dbms/src/Flash/Planner/Plans/PhysicalExpand.cpp

```diff
@@ -0,0 +1,132 @@
+#include <Flash/Pipeline/Exec/PipelineExecBuilder.h>
+#include <Flash/Planner/FinalizeHelper.h>
+#include <Flash/Planner/PhysicalPlanHelper.h>
+#include <Flash/Planner/Plans/PhysicalExpand.h>
+#include <Interpreters/Context.h>
+#include <Operators/ExpressionTransformOp.h>
+#include <Operators/FilterTransformOp.h>
+#include <fmt/format.h>
+
+namespace DB
+{
+PhysicalPlanNodePtr PhysicalExpand::build(
+    const Context & context,
+    const String & executor_id,
+    const LoggerPtr & log,
+    const tipb::Expand & expand,
+    const PhysicalPlanNodePtr & child)
+{
+    assert(child);
+
+    child->finalize();
+
+    if (unlikely(expand.grouping_sets().empty()))
+    {
+        //should not reach here
+        throw TiFlashException("Expand executor without grouping sets", Errors::Planner::BadRequest);
+    }
+
+    DAGExpressionAnalyzer analyzer{child->getSchema(), context};
+    ExpressionActionsPtr before_expand_actions = PhysicalPlanHelper::newActions(child->getSampleBlock());
+    ExpressionActionsPtr expand_actions_itself = PhysicalPlanHelper::newActions(child->getSampleBlock());
+
+    auto grouping_sets = analyzer.buildExpandGroupingColumns(expand, before_expand_actions);
+    auto expand_action = ExpressionAction::expandSource(grouping_sets);
+    expand_actions_itself->add(expand_action);
+
+    // construct sample block.
+    NamesAndTypes expand_output_columns;
+    auto child_header = child->getSchema();
+    for (const auto & one : child_header)
+    {
+        expand_output_columns.emplace_back(one.name, expand_action.expand->isInGroupSetColumn(one.name) ? makeNullable(one.type) : one.type);
+    }
+    expand_output_columns.emplace_back(expand_action.expand->grouping_identifier_column_name, expand_action.expand->grouping_identifier_column_type);
+
+    auto physical_expand = std::make_shared<PhysicalExpand>(
+        executor_id,
+        expand_output_columns,
+        log->identifier(),
+        child,
+        expand_action.expand,
+        expand_actions_itself,
+        Block(expand_output_columns));
+
+    return physical_expand;
+}
+
+
+void PhysicalExpand::expandTransform(DAGPipeline & child_pipeline)
+{
+    String expand_extra_info = fmt::format("expand, expand_executor_id = {}", execId());
+    FmtBuffer fb;
+    fb.append(": grouping set ");
+    shared_expand->getGroupingSetsDes(fb);
+    expand_extra_info.append(fb.toString());
+    child_pipeline.transform([&](auto & stream) {
+        stream = std::make_shared<ExpressionBlockInputStream>(stream, expand_actions, log->identifier());
+        stream->setExtraInfo(expand_extra_info);
+    });
+}
+
+void PhysicalExpand::buildPipelineExec(PipelineExecGroupBuilder & group_builder, Context &, size_t)
+{
+    auto input_header = group_builder.getCurrentHeader();
+    group_builder.transform([&](auto & builder) {
+        builder.appendTransformOp(std::make_unique<ExpressionTransformOp>(group_builder.exec_status, expand_actions, log->identifier()));
```

Comment:
You can modify `Pipeline::isSupported(const tipb::DAGRequest & dag_request)` to enable tests for pipeline model in `gtest_expand_executor.cpp`

Reply:
make sense, great!